Plutarch/List.hs

```diff
@@ -87,27 +90,27 @@ instance PListLike PList where
   pelimList match_cons match_nil =
     plam $ \ls -> pmatch ls $ \case
       PSCons x xs -> match_cons # x # xs
-      PSNil -> pforce match_nil
+      PSNil -> match_nil
   pcons = plam $ \x xs -> pcon (PSCons x xs)
   pnil = pcon PSNil
 
 type PIsListLike list a = (PListLike list, PElemConstraint list a)
 
 -- | / O(n) /. Convert from any ListLike to any ListLike, provided both lists' element constraints are met.
 pconvertLists :: forall f g a s. (PElemConstraint f a, PElemConstraint g a, PListLike f, PListLike g) => Term s (f a :--> g a)
-pconvertLists =
+pconvertLists = phoistAcyclic $
   pfix #$ plam $ \self ->
     pelimList
```

Comment:
curious question: is `pelimList` sufficiently efficient to be using it in most list utilities?

Reply:
Good question. I haven't looked at it too much.

Reply:
Actually, can you change it to be a zero-cost wrapper? Right now it does a bunch of unnecessary stuff. This will be a breaking change, but :shrug: it's only been a day. It should preferably incur no overhead compared to doing `pmatch`.

Reply:
Do you mean something like (on the `PList` instance, for example)-
```hs
pelimList :: PElemConstraint list a => Term s (a :--> list a :--> r) -> Term s r -> Term s (list a) -> Term s r
pelimList match_cons match_nil ls = pmatch ls $ \case
    PSCons x xs -> match_cons # x # xs
    PSNil -> match_nil
```

I think we would ideally want to preserve the ability to take in a `match_cons` and `match_nil` separately - but I wonder whether or not they should be Haskell level functions.

Reply:
From looking at the usages of `pelimList` around the codebase - my intuition is that a haskell level `match_cons` would probably be better. But I might be wrong.

Reply:
I want to experiment with this a bit alongside a few other changed on the `pelimList-test` branch - I'll use the minswap rewrite to benchmark them.

I'll let you know when I'm happy with the results so you can look into merging this.

Reply:
In general Haskell-level functions are probably better now due to the η-reduction optimisations. We still don't have any good inlining, so `(plam $ \x y -> x y) # x # y` does not reduce to `x y` unfortunately.

Reply:
Yeah, we should have an issue for that.